api/src/test/java/org/apache/druid/timeline/SegmentIdTest.java

```diff
@@ -0,0 +1,179 @@
+  }
+
+  @Test
+  public void testInvalidFormat1()
+  {
+    Assert.assertNull(SegmentId.tryParse("datasource", "datasource_invalid_interval_version"));
+  }
+
+  @Test
+  public void testInvalidFormat2()
+  {
+    Assert.assertNull(SegmentId.tryParse("datasource", "datasource_2015-01-02T00:00:00.000Z_version"));
+  }
+
+  /**
+   * Tests that {@link SegmentId#tryExtractMostProbableDataSource} successfully extracts data sources from some
+   * reasonable segment ids.
+   */
+  @Test
+  public void testTryParseHeuristically()
+  {
+    List<String> segmentIds = Arrays.asList(
+        "datasource_2015-01-02T00:00:00.000Z_2015-01-03T00:00:00.000Z_ver_0_1",
+        "datasource_2015-01-02T00:00:00.000Z_2015-01-03T00:00:00.000Z_ver",
+        "datasource_1_2015-01-02T00:00:00.000Z_2015-01-03T00:00:00.000Z_ver_0_1",
+        "datasource_1_2015-01-02T00:00:00.000Z_2015-01-03T00:00:00.000Z_ver",
+        "datasource_2015-01-01T00:00:00.000Z_2015-01-02T00:00:00.000Z_2015-01-03T00:00:00.000Z_ver_0_1",
+        "datasource_2015-01-01T00:00:00.000Z_2015-01-02T00:00:00.000Z_2015-01-03T00:00:00.000Z_ver"
+    );
+    for (String segmentId : segmentIds) {
+      String dataSource = SegmentId.tryExtractMostProbableDataSource(segmentId);
+      Assert.assertTrue("datasource".equals(dataSource) || "datasource_1".equals(dataSource));
+      Assert.assertTrue(!SegmentId.iteratePossibleParsingsWithDataSource(dataSource, segmentId).isEmpty());
+    }
+  }
+
+  @Test
+  public void testTryParseVersionAmbiguity()
+  {
+    SegmentId segmentId =
+        SegmentId.tryParse("datasource", "datasource_2015-01-02T00:00:00.000Z_2015-01-03T00:00:00.000Z_ver_0");
+    Assert.assertNotNull(segmentId);
+    Assert.assertEquals("ver_0", segmentId.getVersion());
+    Assert.assertEquals(0, segmentId.getPartitionNum());
+  }
+
+  @Test
+  public void testIterateAllPossibleParsings()
+  {
+    String segmentId = "datasource_2015-01-01T00:00:00.000Z_2015-01-02T00:00:00.000Z_2015-01-03T00:00:00.000Z_ver_0_1";
+    List<SegmentId> possibleParsings = ImmutableList.copyOf(SegmentId.iterateAllPossibleParsings(segmentId));
+    DateTime dt1 = DateTimes.of("2015-01-01T00:00:00.000Z");
+    DateTime dt2 = DateTimes.of("2015-01-02T00:00:00.000Z");
+    DateTime dt3 = DateTimes.of("2015-01-03T00:00:00.000Z");
+    Set<SegmentId> expected = ImmutableSet.of(
+        SegmentId.of("datasource", new Interval(dt1, dt2), "2015-01-03T00:00:00.000Z_ver_0", 1),
+        SegmentId.of("datasource", new Interval(dt1, dt2), "2015-01-03T00:00:00.000Z_ver_0_1", 0),
+        SegmentId.of("datasource_2015-01-01T00:00:00.000Z", new Interval(dt2, dt3), "ver_0", 1),
+        SegmentId.of("datasource_2015-01-01T00:00:00.000Z", new Interval(dt2, dt3), "ver_0_1", 0)
+    );
+    Assert.assertEquals(4, possibleParsings.size());
+    Assert.assertEquals(expected, ImmutableSet.copyOf(possibleParsings));
+  }
```

Comment:
Could you add tests which checks a timestamp as the version value?

Reply:
Seems that it's tested? Or could you write specifically what test code do you want to be added here?

Reply:
ie 
```java
 SegmentId desc = SegmentId.of(datasource, Intervals.of("2015-01-02/2015-01-03"), "2015-01-02T00:00:00.000Z", 1);
```
Also, how it behaves with double underscore "__".